src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java

```diff
@@ -0,0 +1,235 @@
+        for (String instruction : instructions)
+        {
+            if ("client".equals(instruction))
+            {
+                if (depSide != null)
+                {
+                    throw new DependencyParserException(dep, "Up to one side (client or server) can be specified.");
+                }
+                depSide = Side.CLIENT;
+            }
+            else if ("server".equals(instruction))
+            {
+                if (depSide != null)
+                {
+                    throw new DependencyParserException(dep, "Up to one side (client or server) can be specified.");
+                }
+                depSide = Side.SERVER;
+            }
+            else if ("required".equals(instruction))
+            {
+                if (depRequired)
+                {
+                    throw new DependencyParserException(dep, "'required' can only be specified once.");
+                }
+                if (targetIsAll)
+                {
+                    throw new DependencyParserException(dep, "You can't 'require' everything (*)");
+                }
+                depRequired = true;
+            }
+            else if ("before".equals(instruction) || "after".equals(instruction))
+            {
+                if (depOrder != null)
+                {
+                    throw new DependencyParserException(dep, "'before' or 'after' can only be specified once.");
+                }
+                depOrder = instruction;
+            }
+            else
+            {
+                throw new DependencyParserException(dep, String.format("Found invalid instruction '%s'. Only %s are allowed.", instruction, DEPENDENCY_INSTRUCTIONS));
+            }
+        }
+
+        if (depSide != null && depSide != this.side)
+        {
+            return;
+        }
+
+        ArtifactVersion artifactVersion;
+        try
+        {
+            artifactVersion = VersionParser.parseVersionReference(target);
+        }
+        catch (RuntimeException e)
+        {
+            throw new DependencyParserException(dep, "Could not parse version string.", e);
+        }
+
+        // TODO: enable this in 1.13
+//        if (!targetIsAll)
+//        {
+//            String modId = artifactVersion.getLabel();
+//            sanityCheckModId(dep, modId);
+//        }
+
+        if (!depRequired && depOrder == null && !targetIsBounded)
+        {
+            throw new DependencyParserException(dep, "Soft dependencies must have a version restriction specified.");
+        }
+
+        if (depRequired)
+        {
+            info.requirements.add(artifactVersion);
+        }
+
+        if ("before".equals(depOrder))
+        {
+            info.dependants.add(artifactVersion);
+        }
+        else if ("after".equals(depOrder))
+        {
+            info.dependencies.add(artifactVersion);
+        }
+        else if (!depRequired && depOrder == null)
+        {
+            info.softRequirements.add(artifactVersion);
```

Comment:
This is not strictly speaking a mistake or anything, but if you use the `*` (all) target without any order requirement, you basically have a worthless dependency ("any other mod must be present"). I think we should at least log or document that case.

Reply:
That case is already handled, I'll add some test cases for "all" to make sure.

Reply:
Is it? I only gave it a quick glance and didn't find anything that specifically handled a "soft * dependency". Like I said, it's not the end of the world, but users might forget the instructions part and then wonder why their "all" dependency doesn't do anything.

Reply:
Nevermind, I found the related code ("Soft dependencies must have a version restriction specified"). Derp xD

Reply:
In that case it throws "Soft dependencies must have a version restriction specified."
If you give it a version, it throws "You cannot have a versioned dependency on everything (*)"
I have added the unit tests for all the cases now.